### ISSUE TITLE
Emitter can sometimes send requests to the LRS with a missing statements property

### ISSUE THREAD

**Issue by garemoko** [xAPI-vle/moodle-logstore_xapi#358]
See #357

There should be error handling in the emitter such that regardless of what it receives from the transformer, it won't send a request to the LRS that does not have a statements property.

**Comment by garemoko** [comment]
Confirmed that this issue happens in conjunction with a negative assignment score as described in #357

**Comment by garemoko** [comment]
This looks to be where statements get set to an empty array: https://github.com/xAPI-vle/moodle-logstore_xapi/blob/6e33b639984d4a9e166f9480a7603d6102fcd202/src/transformer/handler.php#L52

But where is that picked up and how does that end up sending an invalid request to the LRS?

**Comment by garemoko** [comment]
src/transformer/handler.php passes the events to src/loader/handler.php which passes them on to the configured loader, e.g. src/loader/lrs.php. 

src/loader/lrs.php passes the events to src/loader/utils/load_in_batches.php which breaks the events up and passes them (in batches) to src/loader/utils/load_batch.php, which extracts the statements from the events here: https://github.com/xAPI-vle/moodle-logstore_xapi/blob/master/src/loader/utils/load_batch.php#L24

From there the statements are sent to the loader function from src/loader/lrs.php which sends the statements to the LRS. 

Somehow if just one of the events in the batch contains no statements, the request to the LRS results in the "Missing required argument: statements" error because it has no request body. This suggests that the $statements property in src/loader/utils/load_batch.php end up being null or something, but without   throwing an error before it sends the statements to the LRS. 

Things to check:
1. What is the value of ```$transformedevent['statements']``` in src/loader/utils/load_batch.php in this case?
2. What is the result of the array_merge immediately following?

**Comment by garemoko** [comment]
Interesting. In fact an empty array of statements does not get generated. Instead, we get a statement with a scaled score of ```-INF```. 

So why does that generate the "Missing required argument: statements" error?

**Comment by garemoko** [comment]
Ok. So the issue is that when we go to JSON encode the statements, PHP returns false because it can't json encode INF. 

I'm thinking we should have some error handling so that if we can't encode the statements as JSON we don't send them to the LRS. 

In fact, really we should be validating the statements beyond just that they are valid JSON. @ryansmith94 is there any reason why this plugin no longer uses TinCanPHP?

**Comment by ryasmi** [comment]
Hey sorry just catching up here, I think we can make two optimisations quite easily, firstly we can validate the statements before we send them (we already do some validation checking in our checks) and secondly we can not send a HTTP request when there are no statements to send (we might actually already be doing that due to the chunking). We stopped using TinCanPHP to reduce dependencies, use the Moodle Curl to better support proxying, and avoid TinCanPHP wrapping HTTP request errors.

Thanks for closing, I think I've reviewed all of the open PRs now.
